Migration to remove the legacy pulses

qnkhuat opened the pull request:
PR to remove Pulse on FE: https://github.com/metabase/metabase/pull/35843
Epic https://github.com/metabase/metabase/issues/34147

Closes https://github.com/metabase/metabase/issues/32990 as pulse is no longer available

qnkhuat committed:
migration to remove the legacy pulses

qnkhuat commented on test/metabase/db/schema_migrations_test.clj:
```diff
@@ -732,8 +731,46 @@
+                                                                          :parameters      "[]"
+                                                                          :created_at      :%now
+                                                                          :updated_at      :%now}))
+            legacy-pulse-id (first (t2/insert-returning-pks! :pulse {:name            "A leagcy pulse"
```
```suggestion
            legacy-pulse-id (first (t2/insert-returning-pks! :pulse {:name            "A legacy pulse"
```

qnkhuat committed:
Update test/metabase/db/schema_migrations_test.clj

qnkhuat committed:
Merge branch 'master' into remove-pulse-and-fe

calherries commented on test/metabase/db/schema_migrations_test.clj:
```diff
@@ -716,14 +715,14 @@
                                                                 :password (str (random-uuid))}))
-            original-db (t2/query {:datasource data-source} "SELECT * FROM metabase_database")
-            original-collections (t2/query {:datasource data-source}    "SELECT * FROM collection")
+            original-db-names (t2/select-fn-set :name :metabase_database)
```
Thank you for tidying this

calherries commented on test/metabase/db/schema_migrations_test.clj:
```diff
@@ -732,8 +731,46 @@
+                                                                             :parameters "[]"
+                                                                             :created_at :%now
+                                                                             :updated_at :%now}))
+
```
```suggestion
```

calherries reviewed (approved):
Very solid stuff

qnkhuat committed:
FE: remove legacy pulse components (#35843)

qnkhuat committed:
Update test/metabase/db/schema_migrations_test.clj

Co-authored-by: Cal Herries <39073188+calherries@users.noreply.github.com>

qnkhuat committed:
Merge branch 'master' into remove-pulse-and-fe

iethree committed:
Ensure collections are loaded before showing entity copy form (#36329)

metamben committed:
Enable auto commit (#36094)

kdoh committed:
Allow toggling of the dashboard subscription sidebar from the icon (#36244)

* allow for toggling of isSharing state from the icon trigger

* add a test for sidebar toggling

* use a more standard method of toggling the state

* account openDashboardSubscriptions starting with the sidebar open

* switch to using sidebar to check for existence, do it once

* update openPulseSubscription helper to account for new behavior

ranquild committed:
[MLv2] Filters FE integration branch (#34191)

likeshumidity committed:
Update collections.md (#36356)

changed format to bullets and added note about the badge showing up for Questions in official collections on dashboards not in official collections

snoe committed:
Test for #35319 (#36218)

* Test for #35319

Closes #35319

* Address review comment and fix linter

dpsutton committed:
fix flaky audit tests (#36331)

* fix flaky audit tests

somehow (and _only_ in cloverage running) we seem to end up with the
audit collection but no cards in it. Then all of the tests are failing
which check permissions.

```
ERROR in (cannot-edit-audit-content-dashboards-over-api) (metabase/util.cljc:114)
expected: (= "You don't have permissions to do that." (mt/user-http-request :rasta :put 403 (str "dashboard/" (u/the-id dashboard)) {:name "My new title"}))
  actual: java.lang.Exception: Not something with an ID: nil
 at metabase.util$error.invokeStatic (metabase/util.cljc:114)
```

This is failing

```clojure
(deftest cannot-edit-audit-content-dashboards-over-api
  (install-audit-db-if-needed!)
  (let [dashboard (t2/select-one :model/Dashboard :collection_id (:id (perms/default-audit-collection)))]
    (is (= "You don't have permissions to do that."
                                                                   ;; this dashboard is nil
                                                                   ;; and u/the-id errors
           (mt/user-http-request :rasta :put 403 (str "dashboard/" (u/the-id dashboard)) {:name "My new title"})))))
```

So let's ensure that all of the cards, dashboards, and collection exist,
otherwise install them. Previously it was just checking that the
collection existed, and for some reason it was empty

* uncomment audit tests

* remove excess comments + rewrite exists? queries

---------

Co-authored-by: bryan <bryan.maass@gmail.com>

alxnddr committed:
fix backport comment command (#36380)

bshepherdson committed:
[MLv2] Underlying records: use card name as label for nested queries (#36318)

Fixes #35340 by using the card's name if this question's source is a card.

escherize committed:
MB_CLOUD_GATEWAY_IPS is settable from env var (#36257)

* MB_CLOUD_GATEWAY_IPS is settable from env var

it can be set to a comma delimited sequence of ips like: "1.2.3.4,5.6.7.8".

* lint fix

* uncomment cloud-gateway-ips setter

* Make sure there's a value for mb-cloud-gateway-ips

- before testing that it gets printed out

* use .first to call .within on a single value

markbastian committed:
Ensure Metadata is Propagated for Standalone Pulse Cards (#36357)

Pulse cards based on models with user-curated metadata were not showing proper formatting for percentages. This is because the singe `execute-card` function in `metabase.pulse.util` was not adding in the card `:result_metadata` when the card was a model.

The metadata is now properly added.

Additionally, `metabase.pulse.pulse-integration-test` has been refactored for reuse. The `with-metadata-data-cards` macro provides 3 examples of cards that should be tested (a base card, a model card, and a derived question) and a function that executeds the pulse and returns the resulting data in the last column (The one we are asserting on).

Note that as we develop the test scenario in this ns there may be further refactoring opportunities, such as returning the attachments from the pulse email as well.

qnkhuat committed:
Sync dynamic tables for snowflake (#36121)

calherries committed:
Use `malli=?` instead of `#hawk/malli` etc (#35481)

qnkhuat committed:
Task ns cleanup: use malli and remove log translation (#36394)

oleggromov committed:
Update e2e docs to emphasize running FE for testing (#36352)

calherries committed:
Uploads: Make the display name for the _mb_row_id column "_mb_row_id" (#36361)

iethree committed:
Update node and java versions for v48 release (#36379)

ranquild committed:
Move filter components to area folder (#36393)

sansolovyov committed:
dashboard_tab.entity_id is nullable now, makes drop-entity-ids work (#36401)

resolves #36365

metamben committed:
Add column-name and table-name to fk-filter drill's display-info (#36217)

metamben committed:
Make with-expression-name set the display-name of the clause (#36203)

Co-authored-by: Kamil Mielnik <kamil@kamilmielnik.com>

camsaul committed:
Do not return zoom/pk/fk-details drills for null cell values (#36385)

kamilmielnik committed:
Expressions in AggregationPicker without legacy query (#36031)

* Add expression-clause-for-legacy-expression

Fixes #34830.

* Add legacy-expression-for-expression-clause

* Fix wrapper function name

* Include ExpressionClause in Aggregatable type

* Use Aggregatable type in aggregate function

* Rename query to legacyQuery in Expressions (part 1)

* Rename query to legacyQuery in Expressions (part 2)

* Rename query to legacyQuery in Expressions (part 3)

* Support MLv2 expressionClause in processSource and ExpressionWidget

* Revert changes to ExpressionStep as it's out of scope of this task

* Use new expressionClause in ExpressionWidget

* Use onSelect instead of onSelectLegacy in handleExpressionChange

* Remove onSelectLegacy from AggregationPicker

* Use nullish coalescing operator

* Assert new onChangeExpression argument in test

* Assert new onChangeExpression argument in test

* Update AggregationPicker tests

* Replace toEqual + expect.objectContaining with toMatchObject

* Add stageIndex to setup, use destructuring for onSelect arguments

* Deal with awkward assertions

* Use props in a conservative way

* Replace legacyQuery.database() with query + metadata

* Add a TODO comment

* Add clause prop to AggregationPicker

* Migrate clause name

* Replace 3rd argument in onChangeExpression with onChangeExpressionClause

* Use destructuring

* Allow AggregationClause, drop expression from ExpressionEditorTextfield

* Make withExpressionName work with AggregationClause

* Rename expressionClause to clause

* Update tests with new interface

* Use overloading instead of generics

* Add function body

* Rename operator to clause

* Revert "Use overloading instead of generics"

This reverts commit 3953f85b8f63764cbb09825fa45f52d82efd7b71.

* Fix tests failing due to useSelect usage in AggregationPicker

* Omit aggregation options converting expressions to legacy

Fixes #36120.

* Remove temporary hack

* Get rid of props spread

* Migrate isExpressionEditorInitiallyOpen to MLv2

* Omit aggregation options converting expressions to legacy

Fixes #36120.

* Normalize legacy expressions as MBQL expressions

* Fix isExpressionEditorInitiallyOpen

* Update expressionName signature

* Pass props in ExpressionStep tests in a usual way

* Drop a conditional statement

* Bring back expression prop for backwards-compatibility

* Update ExpressionWidget validation & tests

* Bring back using expression in componentWillReceiveProps

* Bring back removed assertion

* Remove legacyQuery prop from AggregationPicker

* Add a test case for isExpressionEditorInitiallyOpen (which uses Lib.expressionName)

* Fix legacyQuery creation

* Add assertion for expression name

* Fix committing expression with done button

* Improve assertions in tests

* Make with-expression-name set the display-name of the clause

Fixes #36167.

This is needed to be able to set/change the display name of custom aggregation columns too.

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Unskip an e2e test

* Add expression name assertion

* Add expression name assertion

* Use displayInfo instead of displayName
- see https://github.com/metabase/metabase/pull/36203#discussion_r1410496658

* Deprecate displayName in favor of displayInfo
- see https://metaboat.slack.com/archives/C0645JP1W81/p1701341786914219?thread_ts=1701335646.088359&cid=C0645JP1W81

* Use clearer notation

* Format code

* Use displayInfo instead of expressionName in tests

* Add isNamed attribute to ColumnDisplayInfo and ClauseDisplayInfo

* Use displayInfo instead of expressionName in AggregationPicker

* Drop expressionName

* Drop redundant cast

* Simplify isExpressionEditorInitiallyOpen interface

* Use display-info with :named? instead of expression-name

* Improve comment

* Add :named? to the schema of display-info

* Use conditional types for ExpressionWidget prop

* Use getNotebookStep

* Simplify condition

* Simplify assertions

* Use .lastCall

* Remove isNamed attribute from ColumnDisplayInfo

* Use .lastCall

* Fix typing

* Bring back legacyQuery prop to AggregationPicker

* Fix isExpressionEditorInitiallyOpen for new clauses

* Fix unit tests

* Remove wrong assertion

* Fix assertion

* Trigger a change

* Revert "Trigger a change"

This reverts commit 254913dd85d7aa1059ca712e19e4e9f1e901dc8e.

* Fix assertion

---------

Co-authored-by: Tamás Benkő <tamas@metabase.com>

dragonsahead committed:
Add Metabase version argument to Dockerfile (#36402)

* Add version to Dockerfile

When you want to build with a specific version, you need to pass it on the build args

* Update Dockerfile

iethree committed:
Handle Release Notes for Release Candidates (#36378)

oleggromov committed:
Fix card filter connected to a field not causing an update (#36206)

* Rename DashboardHeaderView to match filename
* Fix non updating cards on filter mapping to a field
* Clean up DashboardApp tests a bit
* Clean up Dashboard.jsx cruft
* Add Dashboard componentDidUpdate "unit tests"
* Simplify setup code as per feedback
* Add test for changing dashboard prop in componentDidMount
* Change constant names
* Fix optional chaining in Dashboard

npfitz committed:
Adding pagination to All Personal Collections page (#36255)

* Adding pagination to All Personal Collections page

* adding unit test

* test adjustments

bshepherdson committed:
[MLv2] Underlying records: show drills on clicks with no `:column` (#36319)

That fixes "chart legend" clicks and "pivot cell" clicks, which should
show underlying records and automatic insights drills.

Fixes #35343 and #35394.

qwef committed:
Filter values leaked in API request on Embedding (#36152)

npfitz committed:
Adding markdown support to search result descriptions (#36360)

* Adding markdown support to search result descriptions

* adding e2e test

jeff-bruemmer committed:
docs - hide when empty (#36427)

markbastian committed:
Updating default `number-formatter` for percents (#36431)

* Updating default `number-formatter` for percents

The number formatter used in some static viz rendering (e.g. tables) in `metabase.pulse.render.common/number-formatter` had logic to render only integer-valued percents for values between 1 and 100 and render two significant digits otherwise.

The FE default formatter uses 2 significant digits consistently in `frontend/src/metabase/lib/formatting/numbers.tsx`:

```js
const PRECISION_NUMBER_FORMATTER = d3.format(".2f");
```

This change modifies `number-formatter` to consistenly use 2 significant digits for percentage types.

Note that this does NOT apply to fields with custom formatting. Presently, custom column formatting does not appear to be applied for HTML tables at all for BE rendering while it is applied on the FE. This is a separate issue. This PR just brings the baseline behaviors based on typing into alignment.

* Fixing formatting test

ranquild committed:
Fix scrolling in filter picker (#36375)

camsaul committed:
Return `:pk` drills for non-PK values in rows with multiple PKs (#36440)

camsaul committed:
Return row count: 2 instead of negative row count (#36389)

qnkhuat committed:
FE: Things to do after pulses are removed (#36407)

WiNloSt committed:
Fix click behavior link to dashboards/questions not using client router (#35975)

* Fix click behavior link to dashboards/questions not using client router

* Make the code cleaner

* Fix unit tests

* Make API clearer

* Address review: Handle case insensitive subpath

* Add E2E tests

* Fix failed E2E tests

* Fix another fail E2E test

* Add a test to ensure invalid URL won't break navigation

* Add tests + ensure multiple level subpath works

levex committed:
team: add levex to BEC (#36434)

deniskaber committed:
Set embedding visual settings before initial render (#36095)

* Move embedding settings prep before rendering

* Revert not needed change

* Add dynamic embedding settings for preview mode, add e2e tests for embedding settings

* Refactor test, remove not needed code

* Updates for e2e test based on code review

kamilmielnik committed:
Expressions in FilterPicker without legacy query (#36396)

* Add expression-clause-for-legacy-expression

Fixes #34830.

* Add legacy-expression-for-expression-clause

* Fix wrapper function name

* Include ExpressionClause in Aggregatable type

* Use Aggregatable type in aggregate function

* Rename query to legacyQuery in Expressions (part 1)

* Rename query to legacyQuery in Expressions (part 2)

* Rename query to legacyQuery in Expressions (part 3)

* Support MLv2 expressionClause in processSource and ExpressionWidget

* Revert changes to ExpressionStep as it's out of scope of this task

* Use new expressionClause in ExpressionWidget

* Use onSelect instead of onSelectLegacy in handleExpressionChange

* Remove onSelectLegacy from AggregationPicker

* Use nullish coalescing operator

* Assert new onChangeExpression argument in test

* Assert new onChangeExpression argument in test

* Update AggregationPicker tests

* Replace toEqual + expect.objectContaining with toMatchObject

* Add stageIndex to setup, use destructuring for onSelect arguments

* Deal with awkward assertions

* Use props in a conservative way

* Replace legacyQuery.database() with query + metadata

* Add a TODO comment

* Add clause prop to AggregationPicker

* Migrate clause name

* Replace 3rd argument in onChangeExpression with onChangeExpressionClause

* Use destructuring

* Allow AggregationClause, drop expression from ExpressionEditorTextfield

* Make withExpressionName work with AggregationClause

* Rename expressionClause to clause

* Update tests with new interface

* Use overloading instead of generics

* Add function body

* Rename operator to clause

* Revert "Use overloading instead of generics"

This reverts commit 3953f85b8f63764cbb09825fa45f52d82efd7b71.

* Fix tests failing due to useSelect usage in AggregationPicker

* Omit aggregation options converting expressions to legacy

Fixes #36120.

* Remove temporary hack

* Get rid of props spread

* Migrate isExpressionEditorInitiallyOpen to MLv2

* Omit aggregation options converting expressions to legacy

Fixes #36120.

* Normalize legacy expressions as MBQL expressions

* Fix isExpressionEditorInitiallyOpen

* Update expressionName signature

* Pass props in ExpressionStep tests in a usual way

* Drop a conditional statement

* Bring back expression prop for backwards-compatibility

* Update ExpressionWidget validation & tests

* Bring back using expression in componentWillReceiveProps

* Bring back removed assertion

* Remove legacyQuery prop from AggregationPicker

* Add a test case for isExpressionEditorInitiallyOpen (which uses Lib.expressionName)

* Fix legacyQuery creation

* Add assertion for expression name

* Fix committing expression with done button

* Improve assertions in tests

* Make with-expression-name set the display-name of the clause

Fixes #36167.

This is needed to be able to set/change the display name of custom aggregation columns too.

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Add expression name assertion

* Unskip an e2e test

* Add expression name assertion

* Add expression name assertion

* Use displayInfo instead of displayName
- see https://github.com/metabase/metabase/pull/36203#discussion_r1410496658

* Deprecate displayName in favor of displayInfo
- see https://metaboat.slack.com/archives/C0645JP1W81/p1701341786914219?thread_ts=1701335646.088359&cid=C0645JP1W81

* Use clearer notation

* Format code

* Use displayInfo instead of expressionName in tests

* Add isNamed attribute to ColumnDisplayInfo and ClauseDisplayInfo

* Use displayInfo instead of expressionName in AggregationPicker

* Drop expressionName

* Drop redundant cast

* Simplify isExpressionEditorInitiallyOpen interface

* Use display-info with :named? instead of expression-name

* Improve comment

* Add :named? to the schema of display-info

* Use conditional types for ExpressionWidget prop

* Use getNotebookStep

* Simplify condition

* Simplify assertions

* Use .lastCall

* Remove isNamed attribute from ColumnDisplayInfo

* Use .lastCall

* Fix typing

* Bring back legacyQuery prop to AggregationPicker

* Fix isExpressionEditorInitiallyOpen for new clauses

* Fix unit tests

* Remove wrong assertion

* Fix assertion

* Trigger a change

* Revert "Trigger a change"

This reverts commit 254913dd85d7aa1059ca712e19e4e9f1e901dc8e.

* Fix assertion

* Allow FilterClause

* Drop legacyFilter usage from FilterPicker, FilterStep & QuestionFilters

* Update FilterPicker unit tests

* Remove unused function

* Fix types in unit test

* Remove redundant waitFor

* Update ColumnDisplayInfo type

* Revert changes to ColumnDisplayInfo, avoid using semanticType & effectiveType attributes

* Introduce NamedClause and CustomExpressionClause types

* Simplify assertions

* Remove casting from FilterPicker

* Rename NamedClause to CustomNameClause

* Exclude AggregationClause from FilterPicker's interfaces

* Avoid casting using generics

ranquild committed:
Move drills FE tests to master (#36454)

sansolovyov committed:
add Snyk workflow (#36349)

Resolves #35677

sansolovyov committed:
put reaction on the comment ASAP when reacting to backport command (#36460)

so that user knows that workflow has started
https://octokit.github.io/rest.js/v20#reactions-create-for-issue-comment

calherries reviewed (approved):
Still looking good

calherries committed:
Move upload-csv! to metabase.upload (#36404)

noahmoss committed:
Ensure that admin perms override group manager in GET /api/user (#36446)

camsaul committed:
Misc zoom-in-bins drill fixes (#36445)

camsaul committed:
quick-filter drill should return contains and does-not-contain for text columns (#36449)

camsaul committed:
Misc `zoom-in-geographic` drill fixes (#36448)

bshepherdson committed:
[MLv2] Implement `automatic-insights` drill (#36443)

Automatic insights drills have some unusual conditions.
This adds `metabase.lib.metadata/editable?` and checks it before
returning any drills.

On settings:
In the app, `MetabaseSettings` is a global singleton and the settings
are sometimes updated in place. In the JS testing environment several
mock settings instances can exist at once, and the global singleton does
not necessarily have the values we want for any given test.

This PR makes `metabase.lib.metadata/setting` check for `describe` and
`it` to see if we're in the testing environment, and to trust the
metadata's `settings` in that case.

camsaul committed:
Misc zoom-in-timeseries drill fixes (#36505)

noahmoss committed:
Fix erroring migration and add test to exercise MB_DB_AUTOMIGRATE=false option (#36387)

sansolovyov committed:
add target branch name to backporting branch for less conflicts (#36479)

gusaiani committed:
Tweak code block in doc (#36501)

kulyk committed:
Fix dashboard filter width in edit mode (#36483)

* Use `minWidth` instead of `width`

* Reserve left padding for the settings icon

metamben committed:
Enable converting simple value expressions from pMBQL to legacy (#36493)

Fixes #36459.

Also makes sure that legacy to pMBQL works and that simple value expressions can be named.

bshepherdson committed:
[MLv2] Unskip FE integration tests for automatic-insights drills (#36529)

This was missed from #36443.

Fixes #33558.

qwef committed:
Throw better error message when passing a resource id as string instead of number (#36491)

markbastian committed:
Apply column formatting to CSV exports (#36484)

* Apply column formatting to CSV exports

This PR applies the same formatting logic to CSV exports as it does to pulse bodies (e.g. HTML).

Formerly, there were two related formatting paths:
- `metabase.query-processor.streaming.common/format-value`, which is from a protocol that takes an object and returns a string. It is what was used to export csv data. It applies no actual formatting, only converts objects to strings.
- `metabase.pulse.render.body/get-format`, which builds a formatter from an optional time zone, column metadata, and visualization settings. This formatter takes a string and formats it. It was only used for rendering inline artifacts, such as embedded HTML in emails.

The first function is insufficient to format row data as it is unaware of the column metadata and viz settings. We need to use that data everywhere data is exported in a uniform way.

The solution is to lift `get-format` from `metabase.pulse.render.body` to a more common location (`metabase.pulse.render.common` in this PR step, but needs to be moved out of the pulse code to be a more shared concern) and use it everywhere artifacts are generated.

For csv export, this was achieved as follows in `metabase.query-processor.streaming.csv`:

```clojure
(defmethod qp.si/streaming-results-writer :csv
  [_ ^OutputStream os]
  (let [writer     (BufferedWriter. (OutputStreamWriter. os StandardCharsets/UTF_8))
        formatters (atom {})]
    (reify qp.si/StreamingResultsWriter
      (begin! [_ {{:keys [ordered-cols results_timezone]} :data} viz-settings]
        (swap! formatters (constantly (zipmap
                                        ordered-cols
                                        (map (fn [col]
                                               (p.common/get-format results_timezone col viz-settings))
                                             ordered-cols))))
        (csv/write-csv writer [(map (some-fn :display_name :name) ordered-cols)])
        (.flush writer))

      (write-row! [_ row _row-num cols {:keys [output-order]}]
        (let [[ordered-row
               ordered-cols] (if output-order
                               (let [row-v  (into [] row)
                                     cols-v (into [] cols)]
                                 [(for [i output-order] (row-v i))
                                  (for [i output-order] (cols-v i))])
                               [row cols])]
          (csv/write-csv writer [(map (fn [c r]
                                        (let [formatter (@formatters c)]
                                          (formatter (common/format-value r))))
                                      ordered-cols ordered-row)])
          (.flush writer)))

      (finish! [_ _]
        ;; TODO -- not sure we need to flush both
        (.flush writer)
        (.flush os)
        (.close writer)))))
```

The formatters for each column are build in the `begin!` function and then looked up in each `write-row!`.  The existing `format-value` is used to produce a string then passed into our looked up column formatter.

Note that the new unit tests simulate a pulse and grab the provided temp files as attachments and analyzes those for correctness. This should work in a CI environment so long as the test process has permission to both write attachments to the temp directory and read those attachments back out. Also note that these tests can be slow (but not terribly so).

Primary changes:
- `metabase.email.messages` - fix spelling
- `metabase.pulse.render.body` - move `get-format` out of this ns
- `metabase.pulse.render.common` - move `get-format` into this ns
- `metabase.query-processor.streaming.csv` - new logic to apply pulse renderer formatting to csv exports
- `metabase.pulse.pulse-integration-test` - adding unit tests

One TODO before a final commit of this PR is to move the `get-format` logic out of a pulse ns into something more general. Ultimately, it would be nice if this was a common capability used by both BE and FE.

* Updated formatter state

The rows in qp.si/StreamingResultsWriter are ordered, so there should be no need to do a lookup by col. We should just be able to map the ordered-formatters (using the same order as the cols) with the ordered row data.

* Updating tests in `metabase.query-processor.streaming-test` to reflect new csv formatting.

* Updated fomatter for tests and consistency

This updates several tests (metabase.query-processor.streaming.csv-test) to reflect the new and correct behavior of using a common csv formatter as well as update the formatting code to correctly handle relational types and datetimes.

* require fmt

* Updating metabase.pulse.render.body-test to reflect `:type/DateTime` formatting.

* Updating sqlite test

* Updating report-timezone-test to reflect new CSV rendering

* Fixing e2e test

snoe committed:
[MLv2] drop-stage-if-empty (#36500)

* [MLv2] drop-stage-if-empty

* Addressing PR comments

npretto committed:
[Help link] feature branch (#36172)

feat:: white labeling option for help link

shaunlebron committed:
Fix settings jwt (#36189)

* fix types
* swap unit test step order

kulyk committed:
Revert "Fix card filter connected to a field not causing an update" (#36534)

This reverts commit 2af32bbc73f25b5f53c4e92a49f92787965cf72d.

uladzimirdev committed:
Upgrade @testing-library/jest-dom to address security issue with its dependency (#36518)

kulyk committed:
Add "refresh_downstream" icon (#36546)

kulyk committed:
Add icon search to Storybook (#36547)

iethree committed:
Use stretch variant on switch (#36503)

lbrdnk committed:
Update server side relative datetime generation honeysql (#36528)

* Use date instead of timestamp

* Update logging and add code for local testing

* Rename datetime to date

markbastian committed:
Refactoring formatting code for common utility (#36490)

* Apply column formatting to CSV exports

This PR applies the same formatting logic to CSV exports as it does to pulse bodies (e.g. HTML).

Formerly, there were two related formatting paths:
- `metabase.query-processor.streaming.common/format-value`, which is from a protocol that takes an object and returns a string. It is what was used to export csv data. It applies no actual formatting, only converts objects to strings.
- `metabase.pulse.render.body/get-format`, which builds a formatter from an optional time zone, column metadata, and visualization settings. This formatter takes a string and formats it. It was only used for rendering inline artifacts, such as embedded HTML in emails.

The first function is insufficient to format row data as it is unaware of the column metadata and viz settings. We need to use that data everywhere data is exported in a uniform way.

The solution is to lift `get-format` from `metabase.pulse.render.body` to a more common location (`metabase.pulse.render.common` in this PR step, but needs to be moved out of the pulse code to be a more shared concern) and use it everywhere artifacts are generated.

For csv export, this was achieved as follows in `metabase.query-processor.streaming.csv`:

```clojure
(defmethod qp.si/streaming-results-writer :csv
  [_ ^OutputStream os]
  (let [writer     (BufferedWriter. (OutputStreamWriter. os StandardCharsets/UTF_8))
        formatters (atom {})]
    (reify qp.si/StreamingResultsWriter
      (begin! [_ {{:keys [ordered-cols results_timezone]} :data} viz-settings]
        (swap! formatters (constantly (zipmap
                                        ordered-cols
                                        (map (fn [col]
                                               (p.common/get-format results_timezone col viz-settings))
                                             ordered-cols))))
        (csv/write-csv writer [(map (some-fn :display_name :name) ordered-cols)])
        (.flush writer))

      (write-row! [_ row _row-num cols {:keys [output-order]}]
        (let [[ordered-row
               ordered-cols] (if output-order
                               (let [row-v  (into [] row)
                                     cols-v (into [] cols)]
                                 [(for [i output-order] (row-v i))
                                  (for [i output-order] (cols-v i))])
                               [row cols])]
          (csv/write-csv writer [(map (fn [c r]
                                        (let [formatter (@formatters c)]
                                          (formatter (common/format-value r))))
                                      ordered-cols ordered-row)])
          (.flush writer)))

      (finish! [_ _]
        ;; TODO -- not sure we need to flush both
        (.flush writer)
        (.flush os)
        (.close writer)))))
```

The formatters for each column are build in the `begin!` function and then looked up in each `write-row!`.  The existing `format-value` is used to produce a string then passed into our looked up column formatter.

Note that the new unit tests simulate a pulse and grab the provided temp files as attachments and analyzes those for correctness. This should work in a CI environment so long as the test process has permission to both write attachments to the temp directory and read those attachments back out. Also note that these tests can be slow (but not terribly so).

Primary changes:
- `metabase.email.messages` - fix spelling
- `metabase.pulse.render.body` - move `get-format` out of this ns
- `metabase.pulse.render.common` - move `get-format` into this ns
- `metabase.query-processor.streaming.csv` - new logic to apply pulse renderer formatting to csv exports
- `metabase.pulse.pulse-integration-test` - adding unit tests

One TODO before a final commit of this PR is to move the `get-format` logic out of a pulse ns into something more general. Ultimately, it would be nice if this was a common capability used by both BE and FE.

* Refactoring formatting code for common utility

This PR refactors `metabase.pulse.render.common` to `metabase.formatter` as this is code we want applied to several locations, not just in pulses. It also updates references to these nses and the consistent alias.

A key observation of this formatting code, and reason for the refactor, is that it is "metabase-aware" in that it takes into account metadata columns and visualization settings when building formatters rather than just being a simple generic date or number formatter. This is a common code path that should be used any time we are rendering static assets and could potentially be used as common FE code with future development.

Moves:
- `metabase.pulse.render.common` to `metabase.formatter`
- `metabase.pulse.render.datetime` to `metabase.formatter.datetime`
- `metabase.pulse.render.common-test` to `metabase.formatter-test`
- `metabase.pulse.render.datetime-test` to `metabase.formatter.datetime-test`

* Ordering consistent aliases in kondo config

* Rebase fix to use formatter ns in streaming.csv

* Adding `metabase.formatter` require.

* Updating require alias on new test.

calherries committed:
Fix BE flake: metabase-enterprise.advanced-permissions.common-test/upload-csv-test (#36475)

qnkhuat committed:
Temporarily skip server-side-relative-datetime-honeysql-form (#36567)

bshepherdson committed:
[MLv2] Fix pivot drills integration with the FE (#36509)

These were broken because `lib.js/pivot-types` was returning CLJS
keywords rather than strings.

Also makes applying drill-thrus less noisy (`log/info` -> `log/debug`).

Unskips the FE integration tests for pivot drills. Fixes #33559.

qnkhuat committed:
Sync remove log translation (#36565)

jeff-bruemmer committed:
docs - pulse removal (#36415)

shaunlebron committed:
Migrate SettingsLdapForm to formik (#36054)

bshepherdson committed:
[MLv2] Only mark selected columns for aggregations with columns (#36572)

If the aggregation is based on `:case` or another non-column
value, errors were being thrown trying to generate refs.

iethree committed:
Upload test results to S3 (#35929)

bshepherdson committed:
[MLv2] Fix warning in CLJS build (#36541)

This was introduced by #36443.

alxnddr committed:
convert user reducer to RTK (#36353)

lbrdnk committed:
Unskip one and skip another relative datetime test (#36580)

* Revert "Temporarily skip server-side-relative-datetime-honeysql-form (#36567)"

This reverts commit 769b5eb4395e37308d654890b791e1354a015735.
This was not flaking.

* Skip test that flakes

dragonsahead committed:
Add title to funnel chart that became a bar chart (#36523)

* Add title to funnel chart that became a bar chart

Simple fix for https://github.com/metabase/metabase/issues/36227

* Fix whitespaces

* Add prettier on the file

* add specs

---------

Co-authored-by: Aleksandr Lesnenko <alxnddr@gmail.com>

johnswanson committed:
Workaround for an H2 bug (#36578)

There's a fun bug in H2: https://github.com/h2database/h2database/issues/3942

To reproduce:

- create a table, then
- create a view based on the table, then
- modify the original table, then
- generate a snapshot

The generated snapshot has the `CREATE VIEW` *before* the `CREATE TABLE`. This results in a view that can't be queried
successfully until it is recompiled. Our workaround is to recompile ALL views immediately after we restore the app DB
from a snapshot.

Commit:
fix some questions with problems re: models (#36597)

metamben committed:
Pretty print #>> and != as operators when possible (#36606)

camsaul committed:
Copy drills dox from #36253 (#36511)

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

camsaul committed:
Use `sorted-set` for `locales.clj` so order doesn't change every time we add a new language (#36617)

kulyk committed:
Extract dashboard `QuestionPicker` from `AddCardSidebar` (#36589)

camsaul committed:
Fix underlying-records drill with nil binned value (#36611)

sansolovyov committed:
understandable error when export target directory is not writable (#36535)

kamilmielnik committed:
Add junit.xml to .gitignore (#36627)

- see https://github.com/metabase/metabase/pull/35929

iethree committed:
Handle latest release check correctly for s3 (#36551)

iethree committed:
Add Finnish language support (#36607)

sansolovyov committed:
fix sending backport error message (#36635)

qnkhuat committed:
Sync index info (#36462)

metamben committed:
Support changing the expression when the type and the name don't change (#36558)

jeff-bruemmer committed:
basic action limitations (#36652)

Commit:
fix some questions with problems re: models (#36653)

iethree committed:
Metabase analytics everywhere (#36605)

bshepherdson committed:
[MLv2] `fk-filter` drills should not appear on native queries (#36660)

They were added incorrectly in fixing #35689.

Fixes #36633.

camsaul committed:
Underlying records drills: preserve the temporal bucket on a breakout column in the previous stage (#36613)

escherize committed:
adding perm-graph filtering on db or group id (#36543)

* a more refined first crack at adding perm-graph access for db or group id

* new routes: filter data-perm-graph on db or group

add tests

* remove inline defs

* code review responses

* fix typo

* fix the other tests

* update the tests to use the right malli schema

* reuse private vars in tests

* pull graph checker into test util ns and apply it

metamben committed:
Make replace-clause use custom expression name if set (#36650)

qnkhuat committed:
Merge branch 'master' into remove-pulse-and-fe

qnkhuat committed:
correct a recent migration comment

qnkhuat committed:
Merge branch 'master' into remove-pulse-and-fe
